contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -3832,108 +3831,75 @@ exec_stmt_change_dbowner(PLtsql_execstate *estate, PLtsql_stmt_change_dbowner *s
-	else
+	if (schema_name == NULL || strlen(schema_name) == 0 || *schema_name == '\0')
 		ereport(ERROR,
-					(errcode(ERRCODE_WRONG_OBJECT_TYPE),
-					 errmsg("A full-text index for table or indexed view \"%s\" has already been created."
-					 , table_name))); 
-
-	/* The above query will be
-	 * executed using ProcessUtility()
-	 */
-	exec_utility_cmd_helper(query_str);
-
-	return PLTSQL_RC_OK;
-}
-
-static int	
-exec_stmt_dropfulltextindex(PLtsql_execstate *estate, PLtsql_stmt_dropfulltextindex *stmt)
-{
-	const char		*schema_name;
-	Oid 		schemaOid;
-	Oid			relid;
-	char		*index_name;
-	char		*table_name;
-	char		*query_str;
-	bool		index_exists = true;
-
-	table_name = stmt->table_name;
-	schema_name = gen_schema_name_for_fulltext_index((char *)stmt->schema_name);
-	schemaOid = LookupExplicitNamespace(schema_name, true);
+			(errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+				errmsg("schema of table \"%s\" is null",
+					table_name)));
 
 	// check if schema exists
 	if (!OidIsValid(schemaOid))
-			ereport(ERROR,
-				(errcode(ERRCODE_UNDEFINED_SCHEMA),
-					errmsg("schema \"%s\" does not exist",
-						stmt->schema_name)));
+		ereport(ERROR,
+			(errcode(ERRCODE_UNDEFINED_SCHEMA),
+				errmsg("schema \"%s\" does not exist",
+					stmt->schema_name)));
 
 	relid = get_relname_relid((const char *) table_name, schemaOid);
 
 	// check if table exists
 	if (!OidIsValid(relid))
-			ereport(ERROR,
-				(errcode(ERRCODE_UNDEFINED_TABLE),
-					errmsg("relation \"%s\" does not exist",
-						table_name)));
-				
-	index_name = get_index_name(relid);
+		ereport(ERROR,
+			(errcode(ERRCODE_UNDEFINED_TABLE),
+				errmsg("relation \"%s\" does not exist",
+					table_name)));
 
-	// check if fulltext index exists in the table
-	if (!index_name)
-		index_exists = false;
+	// Get the existing fulltext index name
+	old_index_name = get_fulltext_index_name(relid);
 
-	if (index_exists)						
-		query_str = gen_dropfulltextindex_cmds(index_name, schema_name);
+	if (is_create)
+	{
+		column_name = stmt->column_name;
+		index_name = stmt->index_name;
+		if (old_index_name)
+			ereport(ERROR,
+				(errcode(ERRCODE_WRONG_OBJECT_TYPE),
+					errmsg("A full-text index \"%s\" for table or indexed view \"%s\" has already been created.",
+						old_index_name, table_name)));
+		else
+			query_str = gen_createfulltextindex_cmds(table_name, schema_name, column_name, index_name);
+	}
 	else
-		ereport(ERROR,
-					(errcode(ERRCODE_UNDEFINED_OBJECT),
-					 errmsg("Table or indexed view \"%s\" does not have a full-text index or user does not have permission to perform this action."
-					 , table_name))); 
+	{
+		if (old_index_name)					
+			query_str = gen_dropfulltextindex_cmds(old_index_name, schema_name);
+		else
+			ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_OBJECT),
+					errmsg("Table or indexed view \"%s\" does not have a full-text index or user does not have permission to perform this action.",
```

Comment:
The error message mentions permissions - have we tested for cases of users that wouldn't have DROP FULLTEXT INDEX permissions

Reply:
Actually this is a generic sql server error message which is thrown when there is exist no fulltext index to be deleted from the specified table.

Reply:
we should add some test cases for permission check as well, what level of permissions apply at schema, table level etc.